Index: xfwm4-4.12.3/src/client.c
===================================================================
--- xfwm4-4.12.3.orig/src/client.c
+++ xfwm4-4.12.3/src/client.c
@@ -171,15 +171,31 @@ getQubesLabel (Client *c)
     ScreenInfo *screen_info;
     DisplayInfo *display_info;
     Atom atom_label;
+    Atom atom_label_color;
 
     screen_info = c->screen_info;
     display_info = screen_info->display_info;
 
     atom_label = XInternAtom(display_info->dpy, "_QUBES_LABEL", TRUE);
+    atom_label_color = XInternAtom(display_info->dpy, "_QUBES_LABEL_COLOR", TRUE);
 
-    if (atom_label == 0) {
-        c->qubes_label_color = QUBES_LABEL_DOM0;
-    } else {
+    c->qubes_label_color = QUBES_LABEL_DOM0;
+    if (atom_label_color) {
+        Atom actual_type;
+        int actual_format;
+        unsigned long nitems = 0, bytes_left;
+        unsigned char *data = 0;
+
+        if ((XGetWindowProperty (display_info->dpy, c->window, atom_label_color, 0L, 1L,
+                        FALSE, XA_CARDINAL, &actual_type, &actual_format, &nitems,
+                        &bytes_left, &data) == Success))
+        {
+            if (nitems == 1 && actual_format == 32) {
+                c->qubes_label_color = *(long*)data;
+            }
+        }
+    }
+    if (c->qubes_label_color == QUBES_LABEL_DOM0 && atom_label) {
         Atom actual_type;
         int actual_format;
         unsigned long nitems = 0, bytes_left;
