Index: xfwm4-4.12.3/src/mypixmap.c
===================================================================
--- xfwm4-4.12.3.orig/src/mypixmap.c
+++ xfwm4-4.12.3/src/mypixmap.c
@@ -612,7 +612,12 @@ pixbuf_create_from_xpm (gpointer handle,
                     1.0*color->green/0xFFFF,
                     1.0*color->blue/0xFFFF,
                     &h, &s, &v);
-            gtk_hsv_to_rgb(override_h, override_s, v, &new_r, &new_g, &new_b);
+            /* Special case for black label, otherwise it is the same as gray -
+             * see #2120 */
+            if (override_h == 0.0 && override_s == 0.0 && override_v == 0.0)
+                gtk_hsv_to_rgb(override_h, override_s, v * 0.2, &new_r, &new_g, &new_b);
+            else
+                gtk_hsv_to_rgb(override_h, override_s, v, &new_r, &new_g, &new_b);
             color->red = new_r*0xFFFF;
             color->green = new_g*0xFFFF;
             color->blue = new_b*0xFFFF;
